GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs

```diff
@@ -865,7 +867,35 @@ protected override void OnMouseWheel(MouseEventArgs e)
             }
             else
             {
-                base.OnMouseWheel(e);
+                if (_lastMouseWheel.ElapsedMilliseconds > 1500)
+                {
+                    _mouseWheelDelta = 0;
+                }
+
+                int scrollLines = SystemInformation.MouseWheelScrollLines switch
+                {
+                    // Value of -1 indicates the "One screen at a time" mouse option.
+                    -1 => _rowHeight > 0 ? (Height / _rowHeight) : 1,
+                    > 0 => SystemInformation.MouseWheelScrollLines,
+                    _ => 1
+                };
+
+                _mouseWheelDelta += scrollLines * e.Delta;
+
+                const int wheelDeltaPerRow = 120;
+
+                if (Math.Abs(_mouseWheelDelta) >= wheelDeltaPerRow)
+                {
+                    int consumedWheelDelta = _mouseWheelDelta - (_mouseWheelDelta % wheelDeltaPerRow);
+                    _mouseWheelDelta -= consumedWheelDelta;
+                    int rowDelta = -consumedWheelDelta / wheelDeltaPerRow;
+                    int toIndex = Math.Clamp(FirstDisplayedScrollingRowIndex + rowDelta, 0, _revisionGraph.Count - 1);
+
+                    // This will raise the Scroll event.
+                    FirstDisplayedScrollingRowIndex = toIndex;
```

Comment:
I just noticed that Math.Clamp() will throw if _revisionGraph.Count == 0. Does this ever happen? Maybe for an instant between operations when the grid is cleared and reloaded? I'd like to suggest the following safeguard (the grid is scrollable if there are at least two rows):

```suggestion
                    int maxRowIndex = _revisionGraph.Count - 1;
                    if (maxRowIndex > 0)
                    {
                        int toRowIndex = Math.Clamp(FirstDisplayedScrollingRowIndex + rowDelta, 0, maxRowIndex);

                        // This will raise the Scroll event.
                        FirstDisplayedScrollingRowIndex = toRowIndex;
                    }
```